### ISSUE TITLE
extent_v in static_assert not detecting inferred array length with rest argument expansion before use

### ISSUE THREAD

**Issue by mstange** [llvm/llvm-project#150426]
https://godbolt.org/z/c61WhKbaz

The `static_assert` below fails in clang 20.1.0, but passes in gcc 15.1 and msvc 19:

```cpp
#include <stdio.h>
#include <type_traits>

template <int... ArgValues>
struct MyTemplateStruct {
    static constexpr int values[] = {ArgValues...};
};

int main() {
    using ConcreteStruct = MyTemplateStruct<5>;

    // printf("test: %d\n", ConcreteStruct::values[0]);

    // This assertion starts passing if you uncomment the above printf
    static_assert(std::extent_v<decltype(ConcreteStruct::values)> == 1,
                  "should detect one element in values member");

    return 0;
}
```

Error message with clang:

```
<source>:15:19: error: static assertion failed due to requirement 'std::extent_v<const int[], 0> == 1': should detect one element in values member
   15 |     static_assert(std::extent_v<decltype(ConcreteStruct::values)> == 1,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:15:67: note: expression evaluates to '0 == 1'
   15 |     static_assert(std::extent_v<decltype(ConcreteStruct::values)> == 1,
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
1 error generated.
```

The first template argument to `extent_v` is `const int[]` without the printf, and `const int[1]` with the printf.

**Comment by Aethezz** [comment]
From what I understand, the issue arises because `static constexpr` arrays initialized with pack expansions have an incomplete type (no specified size), which causes `std::extent_v` to fail in detecting their length. The assertion passes when `print`f is used because accessing the array at runtime forces the compiler to fully instantiate and complete the array type, enabling `std::extent_v` to correctly determine its size.

Explicitly declaring the array size as

`static constexpr int values[sizeof...(ArgValues)] = {ArgValues...};`

resolves the problem by making the array type complete upfront, so the code compiles and runs without errors.

I'm just getting started with contributing and was wondering how to approach a fix if needed, I was thinking:

Fix the compiler’s instantiation order or semantic analysis so that the array type is completed early enough—before traits like `std::extent_v` are evaluated—rather than modifying the `std::extent_v` trait itself

Patch `std::extent`: Modify how `extent_v` handles `static constexpr` arrays with pack expansions by explicitly checking for this incomplete state and computing the size as `sizeof...(Args)`. However, this might just patch a symptom and wouldn't help other functions that rely on the array being complete.
